### PR TITLE
File base configuration

### LINKED ISSUE #99: Use a handle instead of MongoID to reference the oauth server
`api_specs` would be easier to read and define if the `oauth_server_id` value was a human readable value such as `api_oauth` than a MongoID such as `ObjectId("5834110db1c3754274e64f0e")`.

### PR DESCRIPTION
**Pull request by italolelis**
## What does this PR do?

This closes #99 and #83 

* Enables configuration based on a set of files. The rule is one file per API definition and one file per OAuth Server definition. At the moment they must reside in separate directories.
* Split the application in two different ports, an administrative port (defaults to `8081`) and proxies port (defaults to `8080`). This way we avoid route collision with the admin routes and also we don't need to load tons of middlewares for the admin routes that are not necessary.
* API Definition and OAuth Server Definition don't depend on an ID anymore, now the name becomes the unique identifier. This works both in MongoDB and file based configurations.
* Added more documentation on how to configure Janus.
* Handled 404 in a more elegant way
* Added a host matcher middleware

This PR does not support BC and therefore needs to bump the major version before merging.

1. All requests to the management API that were based in the ID of the resource now uses the `name`. For instance, instead of `GET /apis/2a28cab5-5d6d-4237-bb0f-927072cc2a9a` use `GET /apis/example`
2. The management API by default now responds on port `8081`. This means that if you try to request `/apis` on port `8080` as it was before it won't work.
3. By default now Janus will try to look for file-based configurations instead of the MongoDB, if you want to make sure you still use mongodb, please ensure that you have your `DATABASE_DSN` set properly.
4. The fields `preserve_host_header`, `strip_listen_path` and `append_listen_path` have been replaced by `preserve_host`, `strip_path` and `append_path`
5. The `APP_NAME` env variable was removed

### CODE REVIEW

File: cmd/janus/main.go [hunk 1]
```diff
@@ -4,82 +4,72 @@ import (
...
-	oAuthServersRepo, err := oauth.NewMongoRepository(accessor.Session.DB(""))
-	if err != nil {
-		log.Panic(err)
+	dsnURL, err := url.Parse(globalConfig.Database.DSN)
+
+	switch dsnURL.Scheme {
+	case "mongodb":
+		log.Debugf("Trying to connect to %s", globalConfig.Database.DSN)
+		session, err := mgo.Dial(globalConfig.Database.DSN)
+		defer session.Close()
```
Comment (vgarvardt) [inline cmd/janus/main.go]: session can be `nil` in some error cases, so defer must be after err check if err is `nil`

File: cmd/janus/main.go [hunk 2]
```diff
@@ -4,82 +4,72 @@ import (
...
-
-	// create the proxy
-	oAuthServersRepo, err := oauth.NewMongoRepository(accessor.Session.DB(""))
-	if err != nil {
-		log.Panic(err)
+	dsnURL, err := url.Parse(globalConfig.Database.DSN)
+
+	switch dsnURL.Scheme {
+	case "mongodb":
+		log.Debugf("Trying to connect to %s", globalConfig.Database.DSN)
```
Comment (vgarvardt) [inline cmd/janus/main.go]: `log.WithField("dsn", globalConfig.Database.DSN).Debug("Trying to connect to DB")`

File: cmd/janus/main.go [hunk 3]
```diff
@@ -4,82 +4,72 @@ import (
...
+		}
+
+		// create the proxy
+		log.Debug("Loading OAuth servers definitions from Mongo DB")
+		oAuthServersRepo, err = oauth.NewMongoRepository(session)
+		if err != nil {
+			log.Panic(err)
+		}
+	case "file":
+		log.Debug("Loading API definitions from file system")
```
Comment (vgarvardt) [inline cmd/janus/main.go]: `log.WithField("path", dsnURL.Path + "/apis").Debug("Loading API definitions from file system")`

File: cmd/janus/main.go [hunk 4]
```diff
@@ -4,82 +4,72 @@ import (
...
+		repo, err = api.NewFileSystemRepository(dsnURL.Path + "/apis")
+		if err != nil {
+			log.Panic(err)
+		}
+
+		log.Debug("Loading OAuth servers definitions from file system")
```
Comment (vgarvardt) [inline cmd/janus/main.go]: `log.WithField("path", dsnURL.Path + "/auth").Debug("Loading OAuth servers definitions from file system") `

File: examples/apis/example.json [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            "Content-Type"
+        ],
+        "exposed_headers" : [ 
+            "X-Debug-Token", 
+            "X-Debug-Token-Link"
+        ],
+        "enabled" : true
+    },
+    "use_compression" : false
+}
```
Comment (vgarvardt) [inline examples/apis/example.json]: newline

File: examples/apis/posts.json [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            "Content-Type"
+        ],
+        "exposed_headers" : [ 
+            "X-Debug-Token", 
+            "X-Debug-Token-Link"
+        ],
+        "enabled" : true
+    },
+    "use_compression" : false
+}
```
Comment (vgarvardt) [inline examples/apis/posts.json]: newline

File: examples/auth/custom_auth.json [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        "exposed_headers" : [],
+        "enabled" : false
+    },
+    "token_strategy" : {
+        "name" : "jwt",
+        "settings" : {
+            "secret" : "asdasdasdsd"
+        }
+    }
+}
```
Comment (vgarvardt) [inline examples/auth/custom_auth.json]: newline

File: examples/apis/example.json [hunk 8]
```diff
@@ -0,0 +1,50 @@
+{
+    "name" : "Example",
+    "slug" : "example",
```
Comment (vgarvardt) [inline examples/apis/example.json]: what do you think about renaming `slug` to `id` to make it more explicit, as this field is real identification for auth/route now?

File: pkg/api/file_repository.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+		return nil, err
+	}
+
+	for _, f := range files {
+		if strings.Contains(f.Name(), ".json") {
+			filePath := filepath.Join(dir, f.Name())
+			log.Info("Loading API Specification from ", filePath)
+			appConfigBody, err := ioutil.ReadFile(filePath)
+			if err != nil {
+				log.WithError(err).Error("Couldn't load app configuration file")
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: `log.WithError(err).WithField("path", filePath).Error("Couldn't load app configuration file")` - we need to have path in error as log level can be set to `error` on production environment

File: pkg/api/file_repository.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	// Grab json files from directory
+	files, err := ioutil.ReadDir(dir)
+	if nil != err {
+		return nil, err
+	}
+
+	for _, f := range files {
+		if strings.Contains(f.Name(), ".json") {
+			filePath := filepath.Join(dir, f.Name())
+			log.Info("Loading API Specification from ", filePath)
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: `log.WithField("path", filePath).Info("Loading API Specification from file")`

File: pkg/api/file_repository.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+func (r *FileSystemRepository) FindAll() ([]*Definition, error) {
+	var definitions []*Definition
+	for _, definition := range r.definitions {
+		definitions = append(definitions, definition)
+	}
+
+	return definitions, nil
+}
+
+// FindBySlug find a country by the iso2code provided
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: `country` => `Proxy definition`

File: pkg/api/file_repository.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+				log.WithError(err).Error("Can't add the definition to the repository")
+				return nil, err
+			}
+		}
+	}
+
+	return repo, nil
+}
+
+// FindAll fetches all the countries available
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: `country` => `Proxy definition`

File: pkg/api/file_repository.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+	for _, definition := range r.definitions {
+		if definition.Proxy.ListenPath == path {
+			return definition, nil
+		}
+	}
+
+	return nil, ErrAPIDefinitionNotFound
+}
+
+// Add adds a country to the repository
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: `country` => `Proxy definition`

File: pkg/api/file_repository.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+func (r *FileSystemRepository) Add(definition *Definition) error {
+	r.definitions[definition.Slug] = definition
+
+	return nil
+}
+
+// Remove removes a country from the repository
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: `country` => `Proxy definition`

File: pkg/api/file_repository.go [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+func (r *FileSystemRepository) Remove(slug string) error {
+	delete(r.definitions, slug)
+	return nil
+}
+
+func (r *FileSystemRepository) parseDefinition(apiDef []byte) *Definition {
+	appConfig := &Definition{}
+	if err := json.Unmarshal(apiDef, appConfig); err != nil {
+		log.Error("[RPC] --> Couldn't unmarshal api configuration: ", err)
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: `log.WithError(err).Error(...)`

File: pkg/api/file_repository.go [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+			return definition, nil
+		}
+	}
+
+	return nil, ErrAPIDefinitionNotFound
+}
+
+// Add adds a country to the repository
+func (r *FileSystemRepository) Add(definition *Definition) error {
+	r.definitions[definition.Slug] = definition
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: I think we should add sync here to make it thread-safe.

File: pkg/api/file_repository.go [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+	return nil
+}
+
+// Remove removes a country from the repository
+func (r *FileSystemRepository) Remove(slug string) error {
+	delete(r.definitions, slug)
```
Comment (vgarvardt) [inline pkg/api/file_repository.go]: I think we should add sync here to make it thread-safe.

File: pkg/api/mongodb_repository.go [hunk 18]
```diff
@@ -1,81 +1,74 @@
...
 // NewMongoAppRepository creates a mongo country repo
-func NewMongoAppRepository(db *mgo.Database) (*MongoAPISpecRepository, error) {
-	coll := db.C("api_specs")
-
-	return &MongoAPISpecRepository{coll}, nil
+func NewMongoAppRepository(session *mgo.Session) (*MongoAPISpecRepository, error) {
+	return &MongoAPISpecRepository{session}, nil
 }
 
 // FindAll fetches all the countries available
```
Comment (vgarvardt) [inline pkg/api/mongodb_repository.go]: `country` => `Proxy definition`

File: pkg/api/mongodb_repository.go [hunk 19]
```diff
@@ -1,81 +1,74 @@
...
-	err := r.coll.Find(bson.M{"proxy.listen_path": path}).One(&result)
+	session, coll := r.getSession()
+	defer session.Close()
+
+	err := coll.Find(bson.M{"proxy.listen_path": path}).One(&result)
 
 	return result, err
 }
 
 // Add adds a country to the repository
```
Comment (vgarvardt) [inline pkg/api/mongodb_repository.go]: `country` => `Proxy definition`

File: pkg/oauth/file_repository.go [hunk 20]
```diff
@@ -0,0 +1,101 @@
+package oauth
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"net/url"
```
Comment (vgarvardt) [inline pkg/oauth/file_repository.go]: move the line up

File: pkg/oauth/file_repository.go [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+	// Grab json files from directory
+	files, err := ioutil.ReadDir(dir)
+	if nil != err {
+		return nil, err
+	}
+
+	for _, f := range files {
+		if strings.Contains(f.Name(), ".json") {
+			filePath := filepath.Join(dir, f.Name())
+			log.Info("Loading API Specification from ", filePath)
```
Comment (vgarvardt) [inline pkg/oauth/file_repository.go]: `WithField()`

File: pkg/oauth/file_repository.go [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+			appConfigBody, err := ioutil.ReadFile(filePath)
+			if err != nil {
+				log.WithError(err).Error("Couldn't load app configuration file")
```
Comment (vgarvardt) [inline pkg/oauth/file_repository.go]: `WithField("path", filePath)`

File: pkg/oauth/file_repository.go [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+	if false == ok {
+		return nil, ErrOauthServerNotFound
+	}
+
+	return server, nil
+}
+
+// Add adds a country to the repository
+func (r *FileSystemRepository) Add(server *OAuth) error {
+	r.servers[server.Slug] = server
```
Comment (vgarvardt) [inline pkg/oauth/file_repository.go]: I think we should add sync here to make it thread-safe.

File: pkg/oauth/file_repository.go [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+
+	return nil
+}
+
+// Remove removes a country from the repository
+func (r *FileSystemRepository) Remove(slug string) error {
+	delete(r.servers, slug)
```
Comment (vgarvardt) [inline pkg/oauth/file_repository.go]: I think we should add sync here to make it thread-safe.

File: pkg/oauth/mongodb_repository.go [hunk 25]
```diff
@@ -2,72 +2,65 @@ package oauth
...
 }
 
 // NewMongoRepository creates a mongo country repo
-func NewMongoRepository(db *mgo.Database) (*MongoRepository, error) {
-	return &MongoRepository{db.C("oauth_servers")}, nil
+func NewMongoRepository(session *mgo.Session) (*MongoRepository, error) {
+	return &MongoRepository{session}, nil
 }
 
 // FindAll fetches all the countries available
```
Comment (vgarvardt) [inline pkg/oauth/mongodb_repository.go]: `country` => `oauth server`

File: pkg/oauth/mongodb_repository.go [hunk 26]
```diff
@@ -2,72 +2,65 @@ package oauth
...
-		return nil, errors.ErrInvalidID
-	}
-
-	err := r.coll.FindId(bson.ObjectIdHex(id)).One(&result)
+	err := coll.Find(bson.M{"slug": slug}).One(&result)
 
 	return result, err
 }
 
 // Add adds a country to the repository
```
Comment (vgarvardt) [inline pkg/oauth/mongodb_repository.go]: `country` => `oauth server`

File: pkg/oauth/mongodb_repository.go [hunk 27]
```diff
@@ -78,37 +71,45 @@ func (r *MongoRepository) Add(oauth *OAuth) error {
...
-	_, err = r.coll.UpsertId(id, oauth)
+	_, err = coll.Upsert(bson.M{"slug": oauth.Slug}, oauth)
 	if err != nil {
-		log.Errorf("There was an error adding the resource %s", id)
+		log.Errorf("There was an error adding the resource %s", oauth.Name)
 		return err
 	}
 
-	log.Debugf("Resource %s added", id)
+	log.Debugf("Resource %s added", oauth.Name)
```
Comment (vgarvardt) [inline pkg/oauth/mongodb_repository.go]: `log.WithField("name", oauth.Name).Debug("Resource added")`

File: pkg/oauth/mongodb_repository.go [hunk 28]
```diff
@@ -78,37 +71,45 @@ func (r *MongoRepository) Add(oauth *OAuth) error {
...
 	return nil
 }
 
 // Remove removes a country from the repository
```
Comment (vgarvardt) [inline pkg/oauth/mongodb_repository.go]: `country` => `oauth server`

File: pkg/oauth/mongodb_repository.go [hunk 29]
```diff
@@ -78,37 +71,45 @@ func (r *MongoRepository) Add(oauth *OAuth) error {
...
-	err := r.coll.RemoveId(bson.ObjectIdHex(id))
+	err := coll.Remove(bson.M{"slug": slug})
 	if err != nil {
-		log.Errorf("There was an error removing the resource %s", id)
+		log.Errorf("There was an error removing the resource %s", slug)
 		return err
 	}
 
-	log.Debugf("Resource %s removed", id)
+	log.Debugf("Resource %s removed", slug)
```
Comment (vgarvardt) [inline pkg/oauth/mongodb_repository.go]: `WithField(...)`

File: examples/apis/example.json (hunk 8 above)
Comment (italolelis) [reply examples/apis/example.json]: I'm not sure... We could probably do that, but do you think it makes sense to call it ID? We could probably just have the `name` property instead, this way we make sure the name is unique and also it make things simpler for whoever is configuring the API

File: pkg/oauth/file_repository.go (hunk 20 above)
Comment (italolelis) [reply pkg/oauth/file_repository.go]: This is from goimports standard.

**Comment by vgarvardt** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/101/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Pull request merged**